[State Sync] Add bootstrapper component to state sync driver.

JoshLind opened the pull request:
## Motivation

This PR adds the final component to the state sync driver: the bootstrapper. This is the component required to bootstrap a node up to the latest epoch ending ledger info version before it can begin syncing continuously from that point. To achieve this, the PR offers several commits:
1. First, we add several new smoke tests to test that a fullnode can bootstrap correctly. At this point, the tests fail because bootstrapping is not supported.
2. Next, we provide some small cleanups to the state sync utils and ensure errors are thrown correctly whenever we terminate an active data stream.
3. Finally, we add the boostrapper component and plug it into the state sync driver. This is the missing piece of the puzzle that enables the new smoke tests to pass!
4. Simplify some of the unit tests (for now). We'll need to build a bigger framework to start testing the functionality.

One this PR lands, I'll start adding further smoke tests (e.g., for validator functionality) and more comprehensive unit and integration tests to verify edge cases.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

New smoke tests have been added for the new functionality.

## Related PRs

None, but this PR relates to: https://github.com/diem/diem/issues/8906

gregnazario commented on state-sync/state-sync-v2/state-sync-driver/src/utils.rs:
```diff
@@ -58,14 +61,36 @@ pub async fn handle_end_of_stream_or_invalid_payload(
+        _ => NotificationFeedback::PayloadTypeIsIncorrect,
     };
     terminate_stream_with_feedback(
         streaming_service_client,
         data_notification.notification_id,
         notification_feedback,
     )
-    .await
+    .await?;
+
+    // Return an error if the payload was invalid
+    match data_notification.data_payload {
+        DataPayload::EndOfStream => Ok(()),
+        _ => Err(Error::InvalidPayload("Unexpected payload type!".into())),
+    }
```
This is kinda weird cause you could send a response, and get an error back, or get an error back sending even though you would have sent an error at the end anyways.

I'm not sure if it could be cleaner though

JoshLind replied:
I'm not sure I understand the exact concern here. When you say `response`, what do you mean? 🤔 This is basically a helper function to handle a data notification. If it's end of stream, it's fine. But if it's not, it means the data type is unexpected, hence the payload.

gregnazario commented on state-sync/state-sync-v2/state-sync-driver/src/driver.rs:
```diff
@@ -139,6 +150,45 @@ impl<
+                "Received a consensus notification: {:?}",
+                notification
+            )))
+        } else if !self.bootstrapper.is_bootstrapped() {
+            Err(Error::BootstrapNotComplete(format!(
+                "Consensus shouldn't be running! Unable to process notification: {:?}",
+                notification
+            )))
```
Message is a little confusing saying consensus shouldn't be running, making it seem like it's an issue because consensus is ahead.  I think it should probably focus on the fact that state sync isn't ready.  Maybe say something around "Received consensus message, but not yet bootstrapped"?

JoshLind replied:
Done, updated this to "Received consensus notification: {:?}, notification". The error wrapper `BootstrapNotComplete` will mention that state sync hasn't finished bootstrapping.

gregnazario commented on state-sync/state-sync-v2/state-sync-driver/src/bootstrapper.rs:
```diff
@@ -0,0 +1,848 @@
+        self.verified_waypoint
+    }
+
+    /// Sets `verified_waypoint` to true
+    pub fn set_verified_waypoint(&mut self) {
+        self.verified_waypoint = true;
+    }
```
Would this not be set by `verify_waypoint`?  Seems strange to control something like this from the outside.  If it's for testing we should label it appropriately.

JoshLind replied:
Yeah, annoyingly it needs to be accessible/set by two different places: `verify_waypoint` and `verify_waypoint_is_satisfiable()`. The satisfiability check occurs in the bootstrapper and basically sets if it storage has already verified the waypoint in the past (e.g., we just crashed and rebooted), or if the waypoint is genesis.

gregnazario commented on state-sync/state-sync-v2/state-sync-driver/src/bootstrapper.rs:
```diff
@@ -0,0 +1,848 @@
+        let highest_known_ledger_info = highest_known_ledger_info.ledger_info();
+        let highest_local_epoch_end = if highest_known_ledger_info.ends_epoch() {
+            highest_known_ledger_info.epoch()
+        } else if highest_known_ledger_info.epoch() > 0 {
+            highest_known_ledger_info
+                .epoch()
+                .checked_sub(1)
+                .ok_or_else(|| {
+                    Error::IntegerOverflow("The highest local epoch end has overflown!".into())
+                })?
```
FYI, since you check the condition above, it will never fail.

JoshLind replied:
Yeah, I've given up reasoning about which checked arithmetic operations are genuine and those which aren't. I'm basically programming them like a robot now. Once this things all lands, we should figure out the right direction for all of these in our code. 😄

gregnazario reviewed (approved):
I think there are some minor comments, but let's ship it

JoshLind commented:
Thanks @gregnazario! 😄 Note: I added another temporary commit to simplify some of the unit tests that now failed. In order to make them pass, I need to build a better framework (which is the next step anyway). So I've removed some test code and will return to it all when this lands.

gregnazario reviewed (dismissed):
/land
